src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (tail == null) {
+            head = tail = newNode;
+        } else {
+            tail.setNext(newNode);
+            newNode.setPrev(tail);
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+
+        Node<T> newNode = new Node<>(value);
+
+        if (index == 0) {
+            if (head == null) {
+                head = tail = newNode;
+            } else {
+                newNode.setNext(head);
+                head.setPrev(newNode);
+                head = newNode;
+            }
+        } else if (index == size) {
+            add(value);
+            return;
+        } else { // Insert in the middle
+            Node<T> current = getNode(index);
+            newNode.setNext(current);
+            newNode.setPrev(current.getPrev());
+            current.getPrev().setNext(newNode);
+            current.setPrev(newNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            this.add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> node = getNode(index);
+        return node.getData();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
```

Comment:
Create separate method to avoid duplication of
```
if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }
```